keps/sig-storage/20200212-volume-group.md

```diff
@@ -0,0 +1,1067 @@
+Type VolumeGroupSource struct {
+        // A label query over existing persistent volume claims to be added to the volume group.
+        // +optional
+        Selector *metav1.LabelSelector
+
+        // If specified, the VolumeGroup will be created from the source
+        // VolumeGroupSnapshot.
+        // +optional
+        SourceVolumeGroupSnapshotName *string
+}
+
+
+type VolumeGroupStatus struct {
+	// VolumeGroupId is a unique id returned by the CSI driver
+	// to identify the VolumeGroup on the storage system.
+	// If a storage system does not provide such an id, the
+	// CSI driver can choose to return the VolumeGroup name.
+	// +optional
+	VolumeGroupId *string
+
+	// +optional
+        GroupCreationTime *metav1.Time
+
+        // A list of persistent volume claims
+        // +optional
+        PVCList []PersistentVolumeClaim
+
+	// +optional
+        Ready *bool
+
+        // Last error encountered during group creation
+	// +optional
+        Error *VolumeGroupError
+}
+
+// Describes an error encountered on the group
+type VolumeGroupError struct {
+    	// time is the timestamp when the error was encountered.
+    	// +optional
+    	Time *metav1.Time
+ 
+    	// message details the encountered error
+    	// +optional
+    	Message *string
+}
+
+// VolumeGroupSnapshot is a user's request for taking a group snapshot.
+type VolumeGroupSnapshot struct {
+        metav1.TypeMeta `json:",inline"`
+        // Standard object's metadata.
+        // +optional
+        metav1.ObjectMeta `json:"metadata,omitempty" protobuf:"bytes,1,opt,name=metadata"`
+
+        // Spec defines the desired characteristics of a group snapshot requested by a user.
+        Spec VolumeGroupSnapshotSpec `json:"spec" protobuf:"bytes,2,opt,name=spec"`
+
+        // Status represents the latest observed state of the group snapshot
+        // +optional
+        Status *VolumeGroupSnapshotStatus `json:"status,omitempty" protobuf:"bytes,3,opt,name=status"`
+}
+
+// VolumeGroupSnapshotSpec describes the common attributes of a group snapshot
+type VolumeGroupSnapshotSpec struct {
+        // Source has the information about where the group snapshot is created from.
+        // Supported Kind is VolumeGroup
+	// Required.
+        Source TypedLocalObjectReference `json:"source" protobuf:"bytes,1,opt,name=source"`
+}
+
+Type VolumeGroupSnapshotStatus struct {
+        // VolumeGroupSnapshotId is a unique id returned by the CSI driver
+        // to identify the VolumeGroupSnapshot on the storage system.
+        // If a storage system does not provide such an id, the
+        // CSI driver can choose to return the VolumeGroupSnapshot name.
+	// +optional
+	VolumeGroupSnapshotID *string
+
+        // ReadyToUse becomes true when ReadyToUse on all individual snapshots become true
+        // +optional
+        ReadyToUse *bool
+
+        // List of volume snapshots
+	// +optional
+        SnapshotList []VolumeSnapshot
+}
```

Comment:
@humblec Since SnapshotList is optional, I think we can do the following:
- If ReadyToUse is true and SnapshotList is empty, the volume group snapshot controller assumes the storage system does not return individual snapshots.
- If ReadyToUse is true and SnapshotList in not empty, the volume group snapshot controller knows there are individual snapshots created for this group. Those individual snapshots may be used as readonly, but they cannot be removed from the group anyway. 
- In the CSI Spec, this means `repeated .csi.v1.Snapshot snapshots` in VolumeGroupSnapshot message from CreateVolumeGroupSnapshotResponse should be optional, not required.

Reply:
@humblec There's one problem with this though. If the storage system does not know individual snapshots, how do we support restore? Without an individual snapshot, we cannot create a new volume from a snapshot. How do we use this GroupSnapshot?

Reply:
```

@humblec Since SnapshotList is optional, I think we can do the following:

    If ReadyToUse is true and SnapshotList is empty, the volume group snapshot controller assumes the storage system does not return individual snapshots.
    If ReadyToUse is true and SnapshotList in not empty, the volume group snapshot controller knows there are individual snapshots created for this group. Those individual snapshots may be used as readonly, but they cannot be removed from the group anyway.
    In the CSI Spec, this means repeated .csi.v1.Snapshot snapshots in VolumeGroupSnapshot message from CreateVolumeGroupSnapshotResponse should be optional, not required.

```

@xing-yang  The above proposal looks good to me or reasonable, especially when the `snapshotList` is an `optional` parameter . :+1: 

>@humblec There's one problem with this though. If the storage system does not know individual snapshots, how do we support restore? Without an individual snapshot, we cannot create a new volume from a snapshot. How do we use this GroupSnapshot?
>

Exactly, this is where I was bit stuck and tried to list out  the scenario like above.  In a high level it looks like, the entire logic from the controller is derived considering the fact that, the PVCs which are part of a volume group is a `known` identity  for the controller and controller make further decisions like `restore` or `blocking/allowing` an operation on a PVC by looking into the list of PVC objects which are  part of a volume group. I am not sure how much scalable it is, especially when there are many PVCs which are part of one volume group :) , more or less, the volume group and underlying volumes in it becomes a dependency to the controller. I was wondering is there a way to delegate the entire control of a volume group to the storage driver and controller staying away from the underlying objects , rather work on high level object which is volumegroup in this case.

Reply:
I want to understand how your storage system create a volume group and create a group snapshot in this case.
- When you are creating a volume group, does it include a filesystem?
- After you create a group snapshot, how does a user consume the group snapshot?

Reply:
Sure @xing-yang 
> When you are creating a volume group, does it include a filesystem?
>
Yes, the volume group snapshot includes filesystem. 
>After you create a group snapshot, how does a user consume the group snapshot?
>
When the volume group snapshot is taken with some name, only that name is exposed to user. The internal snap names that is maintained at volume level are not exposed currently for the user. ie, snapshot of individual volumes are not user facing rather its kept internally by the storage subsystem ( here cephfs). 

@kotreshhr @batrick anything else to add here from cephfs side.

Reply:
How does a user consume such as snapshot?  Do they mount the filesystem and read it directly?  If that is the case, you don't need to be able to create a PVC from a snapshot.  Therefore you don't need to create another VolumeGroup from this GroupSnapshot.  It is an optional capability.

Reply:
@xing-yang apologies for missing this comment:  let me expand some more on below and try to draw a line:
```

    Case 1: The CreateGroupSnapshot CSI function should return a list of snapshots (Snapshot message defined in CSI Spec) in its response. The GroupSnapshot controller can use the returned list of snapshots to update corresponding individual VolumeSnapshotContents, wait for VolumeSnapshots and VolumeSnapshotContents to be bound, and update SnapshotList in the GroupSnapshot Status.
    Case 2: The CreateGroupSnapshot CSI function returns group_snapshot_id and volume_group_id, but leaves snapshots field as empty. The GroupSnapshot controller watches VolumeSnapshot and VolumeSnapshotContent API objects. When VolumeSnapshot and VolumeSnapshotContent are bound, it saves the VolumeSnapshot API object to SnapshotList in its Status.
```

**On case 2:**  Does the group snapshot controller look for pvc objects in returned `volume_group_id` and check for all `vs` objects and  map these `vs` objects against the PVCList in the returned `volume_group_id` and update the status of `volumegroupsnapshot->snapshotList`?  If this understanding is correct, if we have more than one group snapshot request in flight at the same time, can the controller map the PVCList in returned `volume_group_id` and corresponding `VS` objects properly and update the `Volume Group Snapshot -> Snapshot List`  ?

>A new empty volume group can be created first, and then a new volume can be created from a snapshot one by one and added to the volume group. This can be repeated for all the snapshots in the GroupSnapshot. 
>

While we are on  `restore` with `case 1/2` , afaiu above , the individual restore  with each request carrying VS object from the group snapshot status is the design here.  If yes, does  the controller pass "group_snapshot_id"  as well in these individual  `restore/clone` call  ?

Also, is there a possibility of  single restore call to the CSI driver with `group_snapshot_id` in the request can be part of the design?, so that storage system can do complete volume group restore and return ?    

Please feel free to correct me if I misunderstood the proposal here.

Reply:
> On case 2: Does the group snapshot controller look for pvc objects in returned volume_group_id , check for all vs objects to map these vs objects against the PVCList in the returned volume_group_id and thus update the status of volumegroupsnapshot->snapshotList? If this understanding is correct, if we have more than one group snapshot request in flight at the same time, can the controller map the PVCList in returned volume_group_id and corresponding VS objects properly and update the Volume Group Snapshot -> Snapshot List ?

The group snapshot controller will watch VolumeSnapshots. If a VolumeSnapshot's `volumeGroupSnapshotName` field matches the groupSnapshot name that is being created, it will be added to the SnapshotList of the VolumeGroupSnapshot.

```
apiVersion: snapshot.storage.k8s.io/v1
kind: VolumeSnapshot
metadata:
  name: snapshot1
spec:
  volumeSnapshotClassName: snapClass1
  source:
    persistentVolumeClaimName: pvc1
  volumeGroupSnapshotName: groupSnapshot1
```

Reply:
Updated the KEP.